sql: almost remove context.Context from eval.Context

yuzefovich opened the pull request:
This PR contains several commits that take steps towards the removal of the stored `context.Context` from `eval.Context`. The largest change is introducing an explicit `context.Context` argument to the `tree.TypedExpr.Eval` method (this is done in the first commit). The remaining commits remove accesses to the stored context in most cases (via either already available context in scope or small plumbing).

This PR doesn't completely remove the stored context because there were two methods `eval.Context.MustGetPlaceholderValue` and `eval.UnwrapDatum` (when unwrapping a placeholder) that still use it. Performing the plumbing for those turned out to be a daunting task, so it was abandoned.

In order to not introduce any new usages this PR unexports the stored context, provides the setter method, and marks both as deprecated.

Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85782)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
Is the plan to pass a `context.Context` to `eval.Expr` separately from the `eval.Context`? I sort of feel like we're fighting against a gradient here. I'm worried that this is going to lead to a place where folks think that `eval.Context` can have even longer lifetimes and more flexible uses whereas I'd prefer if we could find a way to work harder to have fewer places where we're allowed to call `eval.Expr` and were more clear on what exactly that `eval.Context`'s state was. Maybe that's wishful thinking.  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-:-N8z_D7A-lDxZy-T7nO8:bnpjz1z)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
> Is the plan to pass a `context.Context` to `eval.Expr` separately from the `eval.Context`?

Yeah, that's my plan. Not sure if I get there though - it's quite daunting of a task.

> I'm worried that this is going to lead to a place where folks think that `eval.Context` can have even longer lifetimes and more flexible uses

Hm, are you implying that the fact that `context.Context` is embedded into `eval.Context` suggests that the latter has shorter lifetime? I feel like most folks would not make such a deduction, so I don't think it'll make things worse than they already are.

IMO storing a `context.Context` is an anti-pattern that needs to have a justification for applying, and in case of `eval.Context` I think it was introduced that way specifically for scalar functions evaluations (which should use the same `context.Context`) and that was ok, but over time we introduced many more callsites out of convenience. Do you agree that removal of `context.Context` from `eval.Context` is good in principle? I believe for scalar functions evaluation we should push the responsibility of supplying the same context on the caller and for all other cases we'll remove the convenience for folks to thing about which context to use too.

I remember Raphael did an [analysis](https://github.com/cockroachdb/cockroach/issues/41992#issuecomment-559928078) of how this monster `eval.Context` struct should be broken down, but that seems even more daunting to do.

ajwerner reviewed (commented):
> suggests that the latter has shorter lifetime? 

I guess my stance is that if we wanted to make it sane for the eval.Context to be a context.Context and embed a context.Context, it ought to have a shorter lifetime than that `context.Context`. It clearly does not. Given that's I'm more or less happy with us removing the `context.Context`. It does leave me wanting to rename `eval.Context` to something else (and, I suppose, `tree.SemaContext`). I don't know what to call it. `eval.Deps`? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-:-N9-3dSz8NZ05NW5A_nP:b-ejkwna)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich committed:
eval: pass context.Context explicitly to Eval

This commit plumbs the `context.Context` argument to `Eval` methods
explicitly in order to remove the context field from `eval.Context` in
the follow-up commits. No additional plumbing is performed, so in many
cases `eval.Context.Context` is passed as this new argument.
Additionally it renames `ctx *eval.Context` to `evalCtx *eval.Context`.
This commit is purely a mechanical change.

Release note: None

yuzefovich committed:
sql: remove most of eval.Context.Context accesses

This commit removes many usages of `eval.Context.Context` in favor of
the newly-plumbed parameter or already available one.

The only change worth calling out is the plumbing of the context for
the optgen-generated `String` implementations - there we use
`context.Background()` since the method is used only for testing and
debugging purposes. This was needed to go around the difficult (and
unnecessary) plumbing into `memo.ScalarFmtInterceptor`.

Release note: None

yuzefovich committed:
sql: remove eval.Context.Ctx()

Release note: None

yuzefovich commented:
@ajwerner I'm curious whether I could nerd-snipe you to take a look at this PR (I'll be asking for a proper review from @msirek), so I'm curious whether either of you could see a way around using `context.TODO` in two spots as mentioned in the last commit.

I started plumbing the `context.Context` as an argument there to address those two spots proper, and it quickly got out of hand (non-rebased [WIP](https://github.com/yuzefovich/cockroach/tree/eval-context-wip) with four extra commits). In there I came to a conclusion that effectively every method of `tree.Datum` interface needed `context.Context` as an argument, and I started going down that route couple months ago but gave up after a few hours.

ajwerner commented:
I don't know if there's a way around it without augmenting a bunch of methods to take a `context.Context` around the datums. I wonder if there's a way to unwind some of this placeholder stuff, but I don't have a good answer. `UnwrapDatum` bothers me on some level. It also seems to me like the root of the problem, on some level, is that all of this comparison stuff is directly in `tree` and doesn't have any `context.Context` parameter. 

What got out of hand when you tried to make the various methods take a context.Context?

yuzefovich commented:
> It also seems to me like the root of the problem, on some level, is that all of this comparison stuff is directly in `tree` and doesn't have any `context.Context` parameter.

Yes, I agree.

> What got out of hand when you tried to make the various methods take a context.Context?

There were just too many callsites (on the order of thousands) of methods of `tree.Datum` interface (and others I had to modify) - I was hoping that only a subset of the methods needed the plumbing (and I finished some), but it was like a sprawling web of more callsites to update, and the more I did the plumbing, the more work was needed to be done ahead :/

ajwerner commented:
I think we need to analyze that web of call sites and understand why the context needs to make its way down through the datums. I have to imagine we've layered something incorrectly -- where some action is lazy which ought to be eager.

yuzefovich commented:
It sounds much easier than doing it in practice 😄 

I think for now I'll drop the last commit that removes the context and, instead, will unexport it and provide setter/getter methods with "deprecated" in the method names. This is so that the commits don't become stale.

yuzefovich committed:
eval: mark stored context.Context as deprecated

This commit unexports the stored `context.Context` from `eval.Context`,
marks it as "deprecated" and provides a setter method for now. There
are just two places that still use it - namely
`eval.Context.MustGetPlaceholderValue` and `eval.UnwrapDatum` (when
unwrapping a placeholder), and refactoring those is extremely difficult
since the amount of plumbing required is probably thousands of line
changes without an easy way to automate that. In order to not lose
progress on the removal achieved in the previous commits we do this
deprecation.

The next step for the complete removal is to do analysis of the web of
callsites of these two methods to see whether we have incorrect
layering, and this will be done separately.

Release note: None

yuzefovich commented:
First commit was mostly an automated replacement, second and third commits will be squashed, fourth and fifth commits will be squashed too. RFAL.

yuzefovich commented:
@cockroach-dev-inf could someone please take a look at `build/bazelutil/nogo_config.json` changes in the last commit?

yuzefovich commented:
@msirek could you please take a look at this? I'd like to merge it before going on PTO (i.e. before Friday).

msirek commented:
> @msirek could you please take a look at this? I'd like to merge it before going on PTO (i.e. before Friday).

Will do. Sorry for the delay.

msirek reviewed (approved):
Nice work!  I just have one question below.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 57 of 122 files at r10, 3 of 224 files at r16, 222 of 222 files at r22, 72 of 72 files at r23, 111 of 111 files at r24, 29 of 29 files at r25, 33 of 33 files at r26, 8 of 10 files at r27, 4 of 4 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-:-NDeXaC83Q1By_VgktzT:bbu8kvu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @msirek and @yuzefovich)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 44 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-NDeUw8cEImIiert70jF:-NDeUw8cEImIiert70jG:b-c4ahli) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f16ef8299e2db090fecd0e56045f8ac70d9443/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L44)):*
> ```Go
> 			panic(pgerror.New(pgcode.InsufficientPrivilege, "permission denied to drop objects"))
> 		}
> 		// TODO: confirm the usage of b as context.Context.
> ```

Is this something that needs to be done before merging?
___
*[`pkg/sql/opt/memo/multiplicity_builder_test.go` line 473 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-NDeUC-yDJQv6YY22gs0:-NDeUC-yDJQv6YY22gs1:b-gscudv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd5cd92556a7b43d9ad0972206d88d172fd294b4/pkg/sql/opt/memo/multiplicity_builder_test.go#L473)):*
> ```Go
> 
> func makeOpBuilder(t *testing.T) testOpBuilder {
> 	ctx := eval.MakeTestingEvalContext(cluster.MakeTestingClusterSettings())
> ```

Rename to `evalCtx`


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

@knz I see that you were writing something - do you have some feedback on this? I want to merge this by the EOD on Thursday if possible.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-:-NDez-9_BGtrGwtFc8S5:bxyvni3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @msirek)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go` line 44 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-NDeUw8cEImIiert70jF:-NDeyum90NPk14NtI9jj:b-ezuy0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/25f16ef8299e2db090fecd0e56045f8ac70d9443/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_owned_by.go#L44)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Is this something that needs to be done before merging?
</blockquote></details>

No, we're good here, removed the TODO, thanks.
___
*[`pkg/sql/opt/memo/multiplicity_builder_test.go` line 473 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/85782#-NDeUC-yDJQv6YY22gs0:-NDekjKM9kzie6hOFmva:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd5cd92556a7b43d9ad0972206d88d172fd294b4/pkg/sql/opt/memo/multiplicity_builder_test.go#L473)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Rename to `evalCtx`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Squashed some commits. Will wait for @knz to respond, and if there are no blocking comments, will merge tomorrow (if CI is green).

yuzefovich commented:
bors r+
